core/common/src/main/java/alluxio/master/PortRegistry.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.net.ServerSocket;
+import java.nio.channels.FileChannel;
+import java.nio.channels.FileLock;
+import java.nio.channels.OverlappingFileLockException;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Registry for reserving ports during tests.
+ *
+ * The registry reserves ports by taking file locks on files in the port coordination directory.
+ * This doesn't prevent external processes from stealing our ports, but it will prevent us from
+ * conflicting with ourselves. We can then run tests in a dockerized environment to completely
+ * prevent conflicts.
+ *
+ * The default coordination directory is determined by the "user.dir" jvm property. The coordination
+ * directory can be overridden by setting the ALLUXIO_PORT_COORDINATION_DIR environment variable.
+ */
+public final class PortRegistry {
+  private static final String PORT_COORDINATION_DIR_PROPERTY = "ALLUXIO_PORT_COORDINATION_DIR";
+
+  public static final PortRegistry INSTANCE = new PortRegistry();
+
+  // Map from port number to the reservation for that port.
+  private final Map<Integer, Reservation> mReserved = new ConcurrentHashMap<>();
+  private final File mCoordinationDir;
+
+  @VisibleForTesting
+  PortRegistry() {
+    String dir = System.getenv(PORT_COORDINATION_DIR_PROPERTY);
+    if (dir == null) {
+      dir = System.getProperty("user.dir");
```

Comment:
If processes are ever started from different working directories, then this strategy wouldn't be able to be used - correct? Otherwise we would need some static location to make it viable regardless of where the process was started.